byterun/pyvm2.py

```diff
@@ -166,18 +167,36 @@ def unwind_block(self, block):
 
     def parse_byte_and_args(self):
         """ Parse 1 - 3 bytes of bytecode into
-        an instruction and optionally arguments."""
+        an instruction and optionally arguments.
+        In Python3.6 the format is 2 bytes per instruction."""
         f = self.frame
         opoffset = f.f_lasti
-        byteCode = byteint(f.f_code.co_code[opoffset])
+        if f.py36_opcodes:
```

Comment:
why do we need special processing here? the non-3.6 branch looks like it should handle this fine.

Reply:
Only until 3.4. I will change the special processing to check for 3.4 onward if that is required.

Reply:
So, using the `opcodes = dis.get_instructions` API is kind of hard in versions before Python 3.6 because on jumps, the VM sets the bytecode counter to the offset to be jumped to. In versions before Python 3.6, there is no direct correspondence to between the opcodes index and the offset to be jumped to -- and has to be computed for each jump. For Python 3.6, and above, there is a direct correspondence because each bytecode is now fixed width, and one can compute the jump index directly by dividing target by 2. Given the significant simplification Python 3.6 has provided, along with the API for bytecode iteration, perhaps we should special case from Python 3.6 onward?

Reply:
@vrthra reports: 

> opcodes = dis.get_instructions API is kind of hard in versions before Python 3.6 because o[f] jumps

While there are more serious problems mentioned below, by using `xdis` in my fork, the newer and better APIs in later relases can be applied to bytecode from an older versions of Python. 

>  the VM sets the bytecode counter to the offset to be jumped to.

In fact, this is wrong at least as far as emulating CPython semantics because this value is stored in the frame's `f_lasti`.

While this may be tolerable for reading  log traces, if you were to try to hook this up with a debugger it would be intolerable: you can't report frame `f_lasti` as the instruction that would be run if the current instruction had finished and not jumped, returned or raised an exception. Instead `f_lasti` needs to show the current instruction as it does in CPython.

> perhaps we should special case from Python 3.6 onward?

Although the specific problems you mention are addressed above, something like this in `x-python` goes on with opcodes.  Here the advantages are

* pedagocial
* cleaner separatation of code, which leads to 
* scalability